feat: use @imgix/gatsby field helpers and support new gatsby-plugin-image v3 API

frederickfogerty opened the pull request:
Hey there Prismic (and @angeloashmore) 👋 

This PR is the start of the process to migrating the Prismic Imgix GraphQL field types over to use @imgix/gatsby internally. A significant benefit of doing this is that the new Gatsby image plugin, gatsby-plugin-image, is supported out of the box.

This PR is not “finished”, but I believe it is in a working state. I’m opening this at an early stage to ensure that I’m going down a path that Prismic is happy with (i.e. in terms of the code) and to get feedback on ways that I could architect the code differently, if the target branch is correct, or if there is any more work to be done (e.g. tests?). @angeloashmore I still have your message about the potential issue you raised on Slack saved, and I’m looking to investigate that this week.

Feel free to review this properly now, or wait a bit longer until I’m fully satisfied that it’s working correctly, and then do a review then. Either way I’m open and would be grateful for feedback!

angeloashmore commented:
Hey @frederickfogerty, thanks for the PR! Really looking forward to getting `gatsby-plugin-image` integrated.

The branches in this project are pretty messy right now, but `alpha` is currently the development branch. Can you target that branch? I'm experimenting with `semantic-release` which is why you see branches like `next`, `beta`, etc.

I'll leave some comments on the changed files. I'm not sure how much you to split work between you and me, so I'm happy to take on anything you feel is out of your scope.

angeloashmore commented on packages/gatsby-source-prismic/src/builders/buildImageBaseFieldConfigMap.ts:
```diff
@@ -1,76 +1,62 @@
 
 const resolveHeight = (source: PrismicAPIImageField): number | undefined =>
-  source.dimensions?.height
+  source.dimensions?.height */
```
The `resolveWidth` and `resolveHeight` functions would be great to have part of creating the fields. Without these functions, the Imgix plugin will need to make a network request for each image to gather width/height metadata to compute the aspect ratio via the `fm=json` Imgix option.

To be more direct of its usage, it could be a single `resolveAspectRatio` function.

`gatsby-plugin-imgix` uses this function to fetch the aspect ratio if the width/height resolve functions return invalid values: https://github.com/WalltoWall/gatsby-plugin-imgix/blob/14592b6eb4678f61f98491e745b477e244adbf4e/src/shared.ts#L54

frederickfogerty replied:
Yep this shouldn't be a problem, I have been working on this in a separate branch and will push this up shortly.

angeloashmore replied:
Awesome!

angeloashmore commented on packages/gatsby-source-prismic/src/builders/buildImageBaseFieldConfigMap.ts:
```diff
@@ -1,76 +1,62 @@
+      ...imgixTypes.enumTypes.map(deps.schema.buildEnumType),
+      ...imgixTypes.inputTypes.map(deps.schema.buildInputObjectType),
+    ]
+    deps.createTypes(types)
```
I've been using `ReaderTaskEither`'s `bind` to help organize everything into `pipe`s. A simple example looks like this: https://github.com/angeloashmore/gatsby-source-prismic/blob/bff33e989d25f8c5ed822efdc9b3778e447c9120/packages/gatsby-source-prismic/src/builders/buildImageThumbnailType.ts#L16

angeloashmore replied:
The base/shared GraphQL types for the plugin are created here: https://github.com/angeloashmore/gatsby-source-prismic/blob/bff33e989d25f8c5ed822efdc9b3778e447c9120/packages/gatsby-source-prismic/src/create-schema-customization.ts#L30-L34

Is there a way to separate the base Imgix types from the fields? This way the types are only created once instead of anytime an image field's type is created (could be many times depending on the repository). Technically it doesn't matter since `graphql-compose` will merge everything, but just figured I'd ask.

frederickfogerty replied:
> Is there a way to separate the base Imgix types from the fields? This way the types are only created once instead of anytime an image field's type is created (could be many times depending on the repository).

Not with the current API design. Technically as long as the same `namespace` is passed every time, all the types will work together (thanks `graphql-compose`), so `createImgixTypes` could be called once in `src/create-schema-customization.ts`, with just the types consumed there and passed to `createTypes`, and then call it again when the fields are needed.

I'm a bit hesitant to refactor this to different functions unless there's a significant improvement to be had, and since I think it's beneficial to have a single API export. What do you think of this?

angeloashmore replied:
That works for me! Creating the base types where the other base types are created in `packages/gatsby-source-prismic/src/create-schema-customization.ts` would be preferrable. The namespace would be consistent so it should work as you described.

I just commented a code change below for `packages/gatsby-source-prismic/src/create-schema-customization.ts`, but if we go this route, it could actually be reverted to the original code. Then `src/builders/buildImgixImageTypes.ts` could be edited instead to return an array of types to create.

angeloashmore commented on packages/gatsby-source-prismic/src/create-schema-customization.ts:
```diff
@@ -1,28 +1,24 @@
+import { createTypes } from './lib/createTypes'
+import { throwError } from './lib/throwError'
 import { Dependencies, Mutable, PluginOptions } from './types'
-import { buildDependencies } from './buildDependencies'
```
Is your editor sorting imports? Is it possible to leave these changes out of the PR?

frederickfogerty replied:
Yeah, my editor sorts imports. I can back these out manually.

angeloashmore replied:
Thank you! This is one of those things where I wish Prettier would automatically sort these things out in a logical way, like Go's formatter.

angeloashmore commented on packages/gatsby-source-prismic/src/create-schema-customization.ts:
```diff
@@ -53,11 +49,10 @@ export const createBaseTypes: RTE.ReaderTaskEither<
     GatsbyGraphQLTypeM.concat(
       scope.baseTypes as Mutable<typeof scope.baseTypes>,
-      scope.imgixTypes,
+      [],
```
If the Imgix types end up not being created here, the whole `RTE.map` call on line 52 can be removed. The `createTypes` call below will need to be modified to accept the `baseTypes` scope value.

(I can handle this if it's not clear what I mean)

frederickfogerty replied:
I did actually try to do this but I wasn't too familiar with `RTE`. 

I understand what you mean but I'm not sure how to do it - could you drop a code suggestion here and I'll accept it?

angeloashmore replied:
Sure thing, it could be rewritten to look like this:

```typescript
export const createBaseTypes: RTE.ReaderTaskEither<
  Dependencies,
  never,
  void
> = pipe(
  [
    buildEmbedType,
    buildGeoPointType,
    buildImageDimensionsType,
    buildImageThumbnailType,
    buildLinkType,
    buildLinkTypeEnumType,
    buildSliceInterface,
    buildStructuredTextType,
    buildTypePathType,
  ],
  RTE.sequenceArray,
  RTE.chain((types) => createTypes(types as Mutable<typeof types>)),
  RTE.map(constVoid),
)
```

Since the array no longer needs to be merged with the Imgix types, it could be simplified into just a list of types being passed into the `createTypes` function. I've found `RTE` to be pretty powerful to work with and organize code, so let me know if you want to learn more!

angeloashmore replied:
I tried doing the code suggestion thing but couldn't get it to edit the rest of the function. Sorry!

frederickfogerty replied:
Thanks for doing this, even though it turns out since we're using `buildImgixImageTypes` again we won't use this (as you said)

frederickfogerty commented:
> The branches in this project are pretty messy right now, but alpha is currently the development branch. Can you target that branch? I'm experimenting with semantic-release which is why you see branches like next, beta, etc.

No problem! Also let me know if I can help with setting up Semantic Release, we use it a lot at imgix.

> I'll leave some comments on the changed files. I'm not sure how much you want to split work between you and me, so I'm happy to take on anything you feel is out of your scope.

Sounds good, I'll make it clear if there's something I want you to take on.

frederickfogerty commented:
Hey @angeloashmore I think this should be all good to go now! I think I've resolved all the pending comments and also handled the default params bug you raised on Slack.

angeloashmore commented:
Thanks @frederickfogerty, this looks great. Everything looks like it should work as expected.

Just a couple of things that would be needed for a full switchover to `@imgix/gatsby` (my apologies, I should have brought this up sooner):

## 1. **gatsby-plugin-prismic-previews**

`gatsby-source-prismic` now has a sibling plugin called `gatsby-plugin-prismic-previews` that handles live CMS previews. Part of the plugin includes reshaping data from Prismic's API to match the GraphQL data from Gatsby, but in the browser. This means `@imgix/gatsby` would need to be added to this plugin as well using the browser-facing builders using `getGatsbyImageData`, `buildFluidImageData`, and `buildFixedImageData`.

You can see the current `gatsby-plugin-imgix` functions being used here: https://github.com/angeloashmore/gatsby-source-prismic/blob/bf9a6977abb01087a858e69a698f4b9258947416/packages/gatsby-plugin-prismic-previews/src/fieldProxies/imageFieldProxy.ts#L58-L77

Are we able to get those integrated as well? In checking the source code for the Fluid and Fixed functions, I'm not sure if it supports gatsby-image placeholders. This is not 100% critical during a preview session, but it could still be beneficial for 100% feature parity. I see `buildImgixFixed` elsewhere in the plugin does support this, so I may be understanding the code incorrectly.

## 2. Tests

The tests in this project require some nuanced knowledge on how the project is setup, so feel free to pass this off to me if you want. I'll list areas where tests need updates for whoever ends up making the adjustments.

**Testing that the shared types are created**

A `gatsbyImageData` field would need to be added just like the `fixed` and `fluid` fields.

https://github.com/angeloashmore/gatsby-source-prismic/blob/bf9a6977abb01087a858e69a698f4b9258947416/packages/gatsby-source-prismic/test/create-schema-customization-image.test.ts#L38-L49

Same here:

https://github.com/angeloashmore/gatsby-source-prismic/blob/bf9a6977abb01087a858e69a698f4b9258947416/packages/gatsby-source-prismic/test/create-schema-customization-image.test.ts#L85-L96

**Testing that the preview data matches what's expected**

Once the `gatsby-plugin-prismic-previews` integrating is done, the following test should be updated to check that the `gatsbyImageData` field is of the correct shape. This is less about testing `@imgix/gatsby`'s output and more about ensuring the the preview plugin is properly adding the field. You can see `sinon` is being used to just check that an object with the correct properties is being provided without caring what the values are.

https://github.com/angeloashmore/gatsby-source-prismic/blob/bf9a6977abb01087a858e69a698f4b9258947416/packages/gatsby-plugin-prismic-previews/test/usePrismicPreviewBootstrap-field-proxies.test.ts#L341-L359

frederickfogerty commented:
Hey there @angeloashmore, thanks for the detailed reply on this!

1. Yep I can definitely export these. I'd be happy to integrate them into the previews but I'm not exactly sure how to go about setting up a test project and so maybe it'd be more efficient if you tried to get them working in a test project you have set up? Would that be okay? I'll work on this today/tomorrow and post some code samples when I have exported them.
2. Thanks for the information but I think since you have familiarity with the tests, as you said, it'd be more efficient for you to go in here quickly and update the tests. Is that okay?

frederickfogerty commented:
Hey @angeloashmore, I implemented and exported the helpers needed for (1) and updated this PR to point to a version that includes them, so I think this should be ready to go!
